File: generation-templates/objective-c-sources/PMEntityManager.m
//
//  PMEntityManager.m
//  essai
//
//  Created by Pierre Molinaro on 28/06/13.
//  Copyright (c) 2013 ECN / IRCCyN. All rights reserved.
//
//-----------------------------------------------------------------------------*

#import "PMAllocationDebug.h"
#import "PMEntityManager.h"
#import "PMManagedEntity.h"
#import "PMRelationshipDescription.h"
#import "easy-bindings-utilities.h"

//-----------------------------------------------------------------------------*

#import <objc/runtime.h>

//-----------------------------------------------------------------------------*

@implementation PMEntityManager

//-----------------------------------------------------------------------------*

- (instancetype) initWithUndoManager: (NSUndoManager *) inUndoManager {
  self = [self init] ;
  if (self) {
    mManagedObjectSet = [NSMutableSet new] ;
    macroAssign (mUndoManager, inUndoManager) ;
    macroNoteObjectAllocation ;
  }
  return self ;
}

//----------------------------------------------------------------------------*
//    Dealloc                                                                 *
//----------------------------------------------------------------------------*

- (void) dealloc {
  macroReleaseSetToNil (mManagedObjectSet) ;
  macroReleaseSetToNil (mUndoManager) ;
  macroNoteObjectDeallocation ;
  macroSuperDealloc ;
}

//-----------------------------------------------------------------------------*

- (NSUndoManager *) undoManager {
  return mUndoManager ;
}

//-----------------------------------------------------------------------------*

- (NSUInteger) entityCount {
  return mManagedObjectSet.count ;
}

//-----------------------------------------------------------------------------*

- (void) addEntity: (PMManagedEntity *) inEntity {
  [self willChangeValueForKey:@"entityCount"] ;
  [mManagedObjectSet addObject:inEntity] ;
  [self didChangeValueForKey:@"entityCount"] ;
  [mUndoManager
    registerUndoWithTarget:self
    selector:@selector (removeEntity:)
    object:inEntity
  ] ;
}

//-----------------------------------------------------------------------------*

- (void) removeEntity: (PMManagedEntity *) inEntity {
  [self willChangeValueForKey:@"entityCount"] ;
  [mManagedObjectSet removeObject:inEntity] ;
  [self didChangeValueForKey:@"entityCount"] ;
  [mUndoManager
    registerUndoWithTarget:self
    selector:@selector (addEntity:)
    object:inEntity
  ] ;
}

//-----------------------------------------------------------------------------*

- (id) newInstanceOfEntity: (Class) inEntityClass {
  PMManagedEntity * result = [inEntityClass alloc] ;
  result = [result initWithEntityManager:self] ;
  [self addEntity:result] ;
  return result ;
}

//-----------------------------------------------------------------------------*

- (void) deleteEntity: (PMManagedEntity *) inObject {
  [inObject resetBeforeDeletion] ;
  [self removeEntity:inObject] ;
}

//-----------------------------------------------------------------------------*

- (void) deleteEntities: (NSArray *) inObjectArray {
  for (PMManagedEntity * object in inObjectArray) {
    [self deleteEntity:object] ;
  }
}

//-----------------------------------------------------------------------------*

- (void) resetBeforeDeletion {
  for (PMManagedEntity * object in mManagedObjectSet) {
    [object resetBeforeDeletion] ;
  }
  [self willChangeValueForKey:@"entityCount"] ;
  macroReleaseSetToNil (mManagedObjectSet) ;
  [self didChangeValueForKey:@"entityCount"] ;
  macroReleaseSetToNil (mUndoManager) ;
}

//-----------------------------------------------------------------------------*

- (NSArray *) allEntitiesKindOfClass: (Class) inClass {
  NSMutableArray * result = [NSMutableArray new] ;
  for (PMManagedEntity * object in mManagedObjectSet) {
    if ([object isKindOfClass:inClass]) {
      [result addObject:object] ;
    }
  }
  macroAutorelease (result) ;
  return result ;
}

//---------------------------------------------------------------------------*
//  R E A C H A B L E   O B J E C T S    F R O M    O B J E C T              *
//---------------------------------------------------------------------------*

- (NSArray *) reachableObjectsFromObject: (PMManagedEntity *) inRootObject {
  NSMutableArray * reachableObjectArray = [NSMutableArray new] ;
  [reachableObjectArray addObject:inRootObject] ;
  NSMutableArray * objectsToExploreArray = [NSMutableArray new] ;
  [objectsToExploreArray addObject:inRootObject] ;
  NSMutableSet * handledObjectSet = [NSMutableSet new] ;
  [handledObjectSet addObject:inRootObject] ;
//---
  while (objectsToExploreArray.count > 0) {
    PMManagedEntity * objectToExplore = [objectsToExploreArray objectAtIndex:0] ;
    [objectsToExploreArray removeObjectAtIndex:0] ;
  //--- To one relation ships
    NSArray * toOneRelationshipNameArray = [objectToExplore toOneRelationshipDescriptionArray] ;
    for (PMRelationshipDescription * description in toOneRelationshipNameArray) {
      id value = [objectToExplore valueForKey:description.relationshipName] ;
      if ((nil != value) && ! [handledObjectSet containsObject:value]) {
        [handledObjectSet addObject:value] ;
        [reachableObjectArray addObject:value] ;
        [objectsToExploreArray addObject:value] ;
      }
    }
  //--- To many relation ships
    NSArray * toManyRelationshipNameArray = [objectToExplore toManyRelationshipDescriptionArray] ;
    for (PMRelationshipDescription * description in toManyRelationshipNameArray) {
      id value = [objectToExplore valueForKey:description.relationshipName] ;
      for (PMManagedEntity * object in value) {
        if (! [handledObjectSet containsObject:object]) {
          [handledObjectSet addObject:object] ;
          [reachableObjectArray addObject:object] ;
          [objectsToExploreArray addObject:object] ;
        }
      }
    }
  }
  macroReleaseSetToNil (handledObjectSet) ;
  macroReleaseSetToNil (objectsToExploreArray) ;
  macroAutorelease
  (reachableObjectArray) ;
  return reachableObjectArray ;
}

//---------------------------------------------------------------------------*
//  R E A C H A B L E   O B J E C T S    F R O M    O B J E C T              *
//---------------------------------------------------------------------------*

- (NSSet *) uneachableObjectsFromObject: (PMManagedEntity *) inRootObject {
  NSArray * reachableObjects = [self reachableObjectsFromObject:inRootObject] ;
  NSMutableSet * s = [NSMutableSet new] ;
  [s setSet:mManagedObjectSet] ;
  [s minusSet:[NSSet setWithArray:reachableObjects]] ;
  macroAutorelease (s) ;
  return s ;
}

//---------------------------------------------------------------------------*
//  S A V I N G    A N D    R E A D I N G                                    *
//---------------------------------------------------------------------------*

- (NSData *) dataForSavingFromRootObject: (PMManagedEntity *) inRootObject {
  NSArray * objectsToSaveArray = [self reachableObjectsFromObject:inRootObject] ;
//--- Set savingIndex for each object
  NSUInteger idx = 0 ;
  for (PMManagedEntity * object in objectsToSaveArray) {
    object.savingIndex = idx ;
    idx ++ ;
  }
//---
  NSMutableArray * saveDataArray = [NSMutableArray new] ;
  for (PMManagedEntity * object in objectsToSaveArray) {
    NSMutableDictionary * d = [NSMutableDictionary new] ; 
    [d setValue:object.className forKey:@"--entity"] ;
    [object saveIntoDictionary:d] ;
    [saveDataArray addObject:d] ;
    macroReleaseSetToNil (d) ;
  }
  macroAutorelease (saveDataArray) ;
  return [NSPropertyListSerialization
    dataFromPropertyList:saveDataArray
    format:NSPropertyListBinaryFormat_v1_0
    errorDescription:nil
  ] ;
}

//-----------------------------------------------------------------------------*

- (id) readFromData: (NSData *) inData
       withRootEntityClass: (Class) inRootEntityClass {
  NSArray * dictionaryArray = [NSPropertyListSerialization
    propertyListFromData:inData
    mutabilityOption:NSPropertyListImmutable 
    format:NULL
    errorDescription:NULL
  ] ;
  NSMutableArray * objectArray = [NSMutableArray new] ;
  for (NSDictionary * d in dictionaryArray) {
    NSString * className = [d objectForKey:@"--entity"] ;
    Class c = objc_getClass ([className cStringUsingEncoding:NSASCIIStringEncoding]) ;
    PMManagedEntity * object = [self newInstanceOfEntity:c] ;
    [objectArray addObject:object] ;
    macroReleaseSetToNil (object) ;
  }
  NSUInteger idx = 0 ;
  for (NSDictionary * d in dictionaryArray) {
    PMManagedEntity * object = [objectArray objectAtIndex:idx] ;
    [object setUpWithDictionary:d withManagedEntityArray:objectArray] ;
    idx ++ ;
  }
  macroAutorelease (objectArray) ;
//--- Return root object
  return [objectArray objectAtIndex:0] ;
}

//-----------------------------------------------------------------------------*

@end

File: 03-simple-document-swift/03-simple-document/PMEntityManager.swift
//
//  PMEntityManager.m
//  essai
//
//  Created by Pierre Molinaro on 28/06/13.
//  Copyright (c) 2013 ECN / IRCCyN. All rights reserved.
//
//-----------------------------------------------------------------------------*

import Cocoa

//-----------------------------------------------------------------------------*

@objc(PMEntityManager) class PMEntityManager : NSObject {
  var mUndoManager : PMUndoManager
  var mManagedObjectSet = NSMutableSet ()

  //-----------------------------------------------------------------------------*
  //  init                                                                       *
  //-----------------------------------------------------------------------------*

  init (undoManager : PMUndoManager) {
    mUndoManager = undoManager
    super.init ()
    noteObjectAllocation (self)
  }

  //-----------------------------------------------------------------------------*
  //  deinit                                                                     *
  //-----------------------------------------------------------------------------*

  deinit {
    noteObjectDeallocation (self)
  }

  //-----------------------------------------------------------------------------*
  //  undoManager                                                                *
  //-----------------------------------------------------------------------------*

  var undoManager : PMUndoManager {
    return mUndoManager
  }

  //-----------------------------------------------------------------------------*
  //  entityCount                                                                *
  //-----------------------------------------------------------------------------*

  func entityCount () -> Int {
    return mManagedObjectSet.count () ;
  }

  //-----------------------------------------------------------------------------*
  //  addEntity                                                                  *
  //-----------------------------------------------------------------------------*

  func addEntity (inEntity : PMManagedEntity) {
    willChangeValueForKey ("entityCount")
    mManagedObjectSet.addObject (inEntity)
    didChangeValueForKey ("entityCount")
    mUndoManager.registerUndoWithTarget (self,
      selector:"private_removeEntity:",
      object:inEntity
    )
  }

  //-----------------------------------------------------------------------------*

  func private_removeEntity (inEntity : PMManagedEntity) {
    willChangeValueForKey ("entityCount")
    mManagedObjectSet.removeObject (inEntity)
    didChangeValueForKey ("entityCount")
    mUndoManager.registerUndoWithTarget (self,
      selector:"addEntity:",
      object:inEntity
    )
  }

  //-----------------------------------------------------------------------------*
  //  newInstanceOfEntityNamed                                                   *
  //-----------------------------------------------------------------------------*

  func newInstanceOfEntityNamed (inEntityTypeName : String) -> PMManagedEntity? {
    var result : PMManagedEntity?
    if inEntityTypeName == "MyRootEntity" {
      result = MyRootEntity (entityManager:self)
    }
    if nil != result {
      addEntity (result!)
    }
    return result
  }

  //-----------------------------------------------------------------------------*
  //  deleteEntity                                                               *
  //-----------------------------------------------------------------------------*

  func deleteEntity (inObject : PMManagedEntity) {
    private_removeEntity (inObject)
  }

  //-----------------------------------------------------------------------------*
  //  deleteEntities                                                             *
  //-----------------------------------------------------------------------------*

  func deleteEntities (inObjectArray : PMManagedEntity []) {
    for object in inObjectArray {
      deleteEntity (object)
    }
  }

  //-----------------------------------------------------------------------------*
  //  allEntitiesKindOfClass                                                     *
  //-----------------------------------------------------------------------------*

  func allEntitiesKindOfClass (inClassName : String) -> PMManagedEntity [] {
    var result : PMManagedEntity [] = []
    for object : AnyObject in mManagedObjectSet {
      let managedObject = object as PMManagedEntity
      if managedObject.className () == inClassName {
        result += managedObject
      }
    }
    return result ;
  }

  //---------------------------------------------------------------------------*
  //  S A V I N G    A N D    R E A D I N G                                    *
  //---------------------------------------------------------------------------*

  func dataForSavingFromRootObject (inRootObject : PMManagedEntity) -> NSData {
    let objectsToSaveArray = reachableObjectsFromObject (inRootObject)
  //--- Set savingIndex for each object
    var idx = 0 ;
    for object in objectsToSaveArray {
      object.savingIndex = idx
      idx = idx + 1
    }
  //---
    var saveDataArray : NSDictionary [] = []
    for object in objectsToSaveArray {
      var d : NSMutableDictionary = [
        "--entity" : object.className ()
      ]
      object.saveIntoDictionary (d)
      saveDataArray += d
    }
    return NSPropertyListSerialization.dataWithPropertyList (saveDataArray,
      format:NSPropertyListFormat.BinaryFormat_v1_0,
      options:0,
      error:nil
    )
  }

  //-----------------------------------------------------------------------------*

  func readFromData (inData : NSData,
                     rootEntityClassName : String) -> PMManagedEntity {
    let v : AnyObject = NSPropertyListSerialization.propertyListWithData (inData,
      options:0, // NSPropertyListReadOptions.Immutable,
      format:nil,
      error:nil
    )
    let dictionaryArray : NSDictionary [] = v as NSDictionary []
    var objectArray : PMManagedEntity [] = []
    for d in dictionaryArray {
      let className = d.objectForKey ("--entity") as String
      let object = newInstanceOfEntityNamed (className)
      objectArray += object!
    }
    var idx = 0
    for d in dictionaryArray {
      var object : PMManagedEntity = objectArray [idx]
      object.setUpWithDictionary (d, managedObjectArray:objectArray)
      idx += 1
    }
  //--- Return root object
    return objectArray [0]
  }

  //---------------------------------------------------------------------------*
  //  R E A C H A B L E   O B J E C T S    F R O M    O B J E C T              *
  //---------------------------------------------------------------------------*

  func reachableObjectsFromObject (inRootObject : PMManagedEntity) -> PMManagedEntity [] {
    var reachableObjectArray : PMManagedEntity [] = [inRootObject]
    var objectsToExploreArray : PMManagedEntity [] = []
    var handledObjectSet = NSMutableSet ()
    handledObjectSet.addObject (inRootObject)
  //---
    while (objectsToExploreArray.count > 0) {
      let objectToExplore = objectsToExploreArray [0]
      objectsToExploreArray.removeAtIndex (0)
    //--- To one relation ships
      let toOneRelationshipNameArray = objectToExplore.toOneRelationshipDescriptionArray ()
      for description in toOneRelationshipNameArray {
        let value : AnyObject = objectToExplore.valueForKey (description.relationshipName)
        if (nil != value) && !handledObjectSet.containsObject (value) {
          handledObjectSet.addObject (value)
          reachableObjectArray += value as PMManagedEntity
          objectsToExploreArray == value as PMManagedEntity
        }
      }
    //--- To many relation ships
      let toManyRelationshipNameArray = objectToExplore.toManyRelationshipDescriptionArray ()
      for description in toManyRelationshipNameArray {
        let value : AnyObject = objectToExplore.valueForKey (description.relationshipName)
        for managedObject in value as PMManagedEntity [] {
          if !handledObjectSet.containsObject (managedObject) {
            handledObjectSet.addObject (managedObject)
            reachableObjectArray += managedObject
            objectsToExploreArray += managedObject
          }
        }
      }
    }
    return reachableObjectArray ;
  }

  //---------------------------------------------------------------------------*
  //  R E A C H A B L E   O B J E C T S    F R O M    O B J E C T              *
  //---------------------------------------------------------------------------*

  func uneachableObjectsFromObject (inRootObject : PMManagedEntity) -> NSSet {
    let reachableObjects = reachableObjectsFromObject (inRootObject)
    var s = NSMutableSet ()
    s.setSet (mManagedObjectSet)
    s.minusSet (NSSet (array:reachableObjects))
    return s
  }

}

//-----------------------------------------------------------------------------*
